### ISSUE TITLE
Bind operator >>= vs >=>

### ISSUE THREAD

**Issue by slang25** [dustinmoris/AspNetCore.Lambda#5]
Not an issue, but question.

Suave took the decision in late 2015 to change its bind operator from `>>=` to `>=>` for [reasons](https://github.com/SuaveIO/suave/pull/335).

Was it a conscious choice to not follow this design?

**Comment by dustinmoris** [comment]
I didn't put too much thought into it to be honest, but I did consider Suave's fish operator and even though I didn't see the thread you linked here I had a feeling it was about avoiding a clash with other libraries.

After thinking about it a moment I thought that the http handlers should probably be only very thin layers where they call into one or a couple of domain functions. Basically by keeping a clear separation of http handlers and domain logic in separate files, you'd naturally never really clash with different `>>=` operators and that's why I guess I just went with the default `>>=` for now. Maybe time will prove me wrong. What do you think?

**Comment by slang25** [comment]
Yeah, makes sense. I was just curious. I cannot think of a real scenario where it would class either to be honest.

**Comment by TheJayMann** [comment]
Also, and I didn't see it mentioned in the referenced issue, but there is a difference between the standard implementations of `(>>=)` and `(>=>)`.  The former is used to apply a monadic value (e.g. `HttpHandlerContext option Async`) to an elevating function (e.g. `HttpHandlerContext -> HttpHandlerContext option Async`), while the latter is used to compose two elevating functions (e.g. `HttpHandlerContext -> HttpHandlerContext option Async`) into a single elevating function.  The operators are similar to the `(|>)` and `(>>)` operators when dealing with regular values (e.g. `HttpHandlerContext`) and regular functions (e.g. `HttpHandlerContext -> HttpHandlerContext`).  Given that your bind implementation is taking two similarly defined functions and combining them into a single similarly defined function, it would make more sense to use the `(>=>)` operator instead of the `(>>=)`, similar to how if it `HttpHandler` was defined as `HttpHandlerContext -> HttpHandlerContext`, one would use the `(>>)` operator rather than the `(|>)` operator.  The proper use of the `(>>=)` operator in this case would have the signature `HttpHandlerContext option Async -> HttpHandler -> HttpHandlerContext option Async`, while the `(>=>)` operator would have the signature of `HttpHandler -> HttpHandler -> HttpHandler`.  More information between the differences of these operators are explained at https://fsharpforfunandprofit.com/posts/elevated-world-2/#bind and https://fsharpforfunandprofit.com/posts/elevated-world-3/#kleisli.

**Comment by dustinmoris** [comment]
Hmm that is interesting. Ok, I'll have a read on those links and re-open this issue meanwhile. I have to say I am still fairly new to F# and also to functional programming (only been working with it for about a year now, any only really intensively in the last 6 months) so I am still learning loads every day and really thankful for good tips like this.

**Comment by slang25** [comment]
This opening post also describes it well: https://github.com/SuaveIO/suave/issues/298#issue-106111101
I would favour `>=>`, but I don't pretend to understand the nuances of all this.

**Comment by dustinmoris** [comment]
Hi, so I read more on this topic and you are definitely right that the current `>>=` is misleading.

I am not sure if I understood it all correctly, but it seems that this would be legit:

```
let compose (handler : HttpHandler) (handler2 : HttpHandler) =
    fun (ctx : HttpHandlerContext) ->
        async {
            let! result = handler ctx
            match result with
            | None      -> return None
            | Some ctx2 ->
                match ctx2.HttpContext.Response.HasStarted with
                | true  -> return  Some ctx2
                | false -> return! handler2 ctx2
        }

let bind (handler : HttpHandler) =
    fun (result : Async<HttpHandlerContext option>) ->
        async {
            let! ctx = result
            match ctx with
            | None -> return None
            | Some c ->
                match c.HttpContext.Response.HasStarted with
                | true  -> return  Some c
                | false -> return! handler c
        }

let (>=>) = compose
let (>>=) = bind
```

Is that correct?

**Comment by TheJayMann** [comment]
That does in fact appear to be correct, though it can be refactored.  Just like how `(>>)` can be defined in terms of a properly defined `(|>)` as `let (>>) f g x = f x |> g`, `(>=>)` can be defined by a properly defined `(>>=)` as `let (>=>) f g x = f x >>= g`.

Given that you are defining named functions and then aliasing them as operators (which is a good idea as it will allow the regular piping operator or function composition operator if someone doesn't want the bind operator or Kleisli operator replaced), your `compose` function can be rewritten as ` let h1 h2 ctx = h1 ctx |> bind h2` or, if you prefer, `let compose h1 h2 ctx = bind h2 (h1 ctx)`. 

Also, if you would rather use the longer parameter names, annotate the parameter types, or pull the ctx into a lambda, that isn't a problem, I simply went for a smaller definition to more easily fit in this reply.

Don't forget that if you implement `compose` in terms of `bind` that you need to move it so that it is defined after bind.

**Comment by slang25** [comment]
Ah, you have both clarified this for me 😃  I understand now. So what about renaming the operator to be a bit more correct?

**Comment by dustinmoris** [comment]
Hi, so yes I will make those changes now and thanks for clarifying all those things. I just finished to migrate the project to .NET SDK RC4, so will pick this up now.

**Comment by dustinmoris** [comment]
Fixed with version [0.1.0-alpha004](https://github.com/dustinmoris/AspNetCore.Lambda/releases/tag/v0.1.0-alpha004).
